File: uva_574.py
data, stack, ans=[], [], []
t=0
def comp(a, b):
    global data
    if len(a)!=len(b):
        return True
    for i in range(len(a)):
        if data[a[i]]!=data[b[i]]:
            return True
    return False

def comb(curi):
    global data, stack, ans, t
    ans_temp=0
    for i in range(len(stack)):
        ans_temp+=data[stack[i]]
    if ans_temp==t:
        check=True
        for item in ans:
            check=comp(item, stack)
            if not check:
                return 0
        if check:
            ans.append(stack)
        return 0
    elif curi>=len(data) or ans_temp>t:
        return 0
    stack.append(curi)
    comb(curi+1)
    stack=stack[:-1]
    comb(curi+1)

while True:
    try:
        input_s=input().split()
    except:
        break
    t, n=int(input_s[0]), int(input_s[1])
    if not t and not n:
        break
    data, stack, ans=[], [], []
    for i in range(2, len(input_s)):
        data.append(int(input_s[i]))
    comb(0)
    print("Sums of {}:".format(t))
    for List in ans:
        for i in range(len(List) ):
            print(data[List[i]], end='')
            if i!=len(List)-1:
                print('+', end='')
        print('')
    if len(ans)==0:
        print('NONE')

File: uva_574.cpp
# include <iostream>
# include <vector>
using namespace std;
int t;
vector<int> data;
vector<int> stack;
vector< vector<int> > ans;
bool comp(vector<int> a, vector<int> b){
    if(a.size()!=b.size())
        return true;
    for (int i=0; i<a.size(); ++i)
        if (data[a[i]]!=data[b[i]])
            return true;
    return false;
}
void comb(int curi){
    int ans_temp=0;
    for (int i=0; i<stack.size(); ++i)
        ans_temp+=data[stack[i]];
    if(ans_temp==t){
        bool check=true;
        for (int i=ans.size()-1; i>=0; --i){
            check=comp(ans[i], stack);
            if (!check)
                return;
        }
        if (check)
            ans.push_back(stack);
        return;
    }
    else if (curi>=data.size() || ans_temp>t)
        return;
    stack.push_back(curi);
    comb(curi+1);
    stack.pop_back();
    comb(curi+1);
}
int main(void){
    int n, temp;
    while(cin >> t >> n){
        if(!t && !n)
            break;
        stack.clear();
        ans.clear();
        data.resize(n);
        for (int i=0; i<n; ++i)
            cin >> data[i];
        comb(0);
        cout << "Sums of " << t << ":\n";
        for (int i=0; i<ans.size(); ++i){
            for (int j=0; j<ans[i].size(); ++j){
                cout << data[ans[i][j]];
                if (j!=ans[i].size()-1)
                    cout << '+';
            }cout << endl;
        }
        if (ans.empty())
            cout << "NONE\n";
    }
    return 0;
}